File: examples/javascript/01_extract.mjs
#!/usr/bin/env node
/**
 * 01_extract — Self-contained extract demo using the JS (koffi) bindings.
 *
 * Creates a DOCX from Markdown (written to a temp file), opens it with
 * Document.open, and prints the format, plain text, and Markdown output.
 * Exit 0 on success.
 *
 * Run: OFFICE_OXIDE_LIB=.../liboffice_oxide.so node examples/javascript/01_extract.mjs
 */
import { tmpdir } from 'os';
import { join } from 'path';
import { unlinkSync } from 'fs';
import { Document, createFromMarkdown } from '../../js/lib/index.js';

const markdown = `# Office Oxide Extract Demo

This document was created from Markdown and parsed back via JavaScript bindings.

## Features

- Plain text extraction
- Markdown conversion
- IR (intermediate representation) access
`;

const tmpPath = join(tmpdir(), `oo_01_extract_${Date.now()}.docx`);

try {
  createFromMarkdown(markdown, 'docx', tmpPath);

  const doc = Document.open(tmpPath);
  try {
    const format = doc.format;
    const text = doc.plainText();
    const md = doc.toMarkdown();
    const ir = doc.toIr();

    console.log('format:', format);
    console.log('--- plain text (first 200 chars) ---');
    console.log(text.slice(0, 200));
    console.log(`--- markdown length: ${md.length} chars ---`);
    console.log(`--- IR sections: ${ir.sections?.length ?? 0} ---`);

    if (format !== 'docx') throw new Error(`expected format=docx, got ${format}`);
    if (!text) throw new Error('plain text is empty');
    if (!text.includes('Office Oxide Extract Demo')) throw new Error('heading missing');

    console.log('\nAll checks passed.');
  } finally {
    doc.close();
  }
} finally {
  try { unlinkSync(tmpPath); } catch (_) {}
}

File: examples/rust/01_extract.rs
//! 01_extract — Self-contained extract demo.
//!
//! Creates a DOCX in memory, reads it back, and prints plain text,
//! Markdown, and an IR summary (number of sections).
//!
//! Run: `cargo run --example 01_extract`

use std::io::Cursor;

use office_oxide::docx::write::DocxWriter;
use office_oxide::{Document, DocumentFormat};

fn main() {
    // ── Build a DOCX in memory ──────────────────────────────────────────────
    let mut writer = DocxWriter::new();
    writer.add_heading("Office Oxide Extract Demo", 1);
    writer.add_paragraph("This document was created in memory and read back via Document::from_reader.");
    writer.add_paragraph("It demonstrates plain-text extraction, Markdown conversion, and IR access.");

    let mut buf = Cursor::new(Vec::new());
    writer.write_to(&mut buf).expect("write DOCX to buffer");
    buf.set_position(0);

    // ── Read it back ────────────────────────────────────────────────────────
    let doc = Document::from_reader(buf, DocumentFormat::Docx).expect("parse DOCX");

    println!("format: {:?}", doc.format());

    let text = doc.plain_text();
    println!("--- plain text ---");
    println!("{text}");

    let md = doc.to_markdown();
    println!("--- markdown ---");
    println!("{md}");

    let ir = doc.to_ir();
    println!("--- IR ---");
    println!("sections: {}", ir.sections.len());
    for (i, s) in ir.sections.iter().enumerate() {
        println!("  section {i}: title={:?}, elements={}", s.title, s.elements.len());
    }

    // ── Verify ─────────────────────────────────────────────────────────────
    assert!(text.contains("Office Oxide Extract Demo"), "heading missing from plain text");
    assert!(text.contains("created in memory"), "paragraph missing from plain text");
    assert!(!ir.sections.is_empty(), "IR must have at least one section");

    println!("\nAll checks passed.");
}
